File: src/core/bbcode/normalize.ts
import { hasChildren, isInternalNode, type InternalNode } from './internal'
import type { BBCodeNode } from './types'

/**
 * 合并相邻文本,并摊平任何没被父标签收走的中间节点。
 *
 * 相邻文本是解析过程的常态:未知标签透传成文本、未闭合标签降级成文本,
 * 都会在已有文本旁边再落一段。渲染层不该关心这些,统一在这里收干净。
 */
export function normalize(nodes: readonly InternalNode[]): BBCodeNode[] {
  const flattened: InternalNode[] = []
  for (const node of nodes) {
    if (isInternalNode(node)) {
      flattened.push(...node.children)
      continue
    }
    flattened.push(node)
  }

  const merged: BBCodeNode[] = []
  for (const node of flattened as BBCodeNode[]) {
    const previous = merged[merged.length - 1]
    if (node.type === 'text' && previous?.type === 'text') {
      merged[merged.length - 1] = { type: 'text', value: previous.value + node.value }
      continue
    }
    merged.push(hasChildren(node) ? { ...node, children: normalize(node.children) } : node)
  }
  return merged
}

/** 取子树里的可见文字,用于只需要一个字符串的节点(如 `[@]名字[/@]`)。 */
export function plainText(nodes: readonly BBCodeNode[]): string {
  return nodes
    .map((node) => {
      if (node.type === 'text') return node.value
      return hasChildren(node) ? plainText(node.children as readonly BBCodeNode[]) : ''
    })
    .join('')
}

File: app/src/main/kotlin/com/chasel/ng2n/core/bbcode/Normalize.kt
package com.chasel.ng2n.core.bbcode

internal fun normalize(nodes: List<ParseNode>): List<BBCodeNode> {
  val flattened = ArrayList<BBCodeNode>(nodes.size)
  flattenInto(nodes, flattened)

  val merged = ArrayList<BBCodeNode>(flattened.size)
  for (node in flattened) {
    val previous = merged.lastOrNull()
    if (node is TextNode && previous is TextNode) {
      merged[merged.size - 1] = TextNode(previous.value + node.value)
      continue
    }
    merged.add(if (node is ChildBearing) node.withChildren(normalize(node.children)) else node)
  }
  return merged
}

private fun flattenInto(nodes: List<ParseNode>, out: MutableList<BBCodeNode>) {
  for (node in nodes) {
    when (node) {
      is InternalNode -> flattenInto(node.children, out)
      is BBCodeNode -> out.add(node)
    }
  }
}

internal fun plainText(nodes: List<BBCodeNode>): String {
  val out = StringBuilder()
  appendPlainText(nodes, out)
  return out.toString()
}

private fun appendPlainText(nodes: List<BBCodeNode>, out: StringBuilder) {
  for (node in nodes) {
    when (node) {
      is TextNode -> out.append(node.value)
      is ChildBearing -> appendPlainText(node.children, out)
      else -> Unit
    }
  }
}
